src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+                                           phantom_node_pair);
+
+            if (!forward_heap.Empty())
+                forward_heap_min = forward_heap.MinKey();
+
+            if (overlap_weight != INVALID_EDGE_WEIGHT && overlap_via != SPECIAL_NODEID)
+            {
+                if (shortest_path_via != SPECIAL_NODEID)
+                {
+                    candidate_vias.push_back(WeightedViaNode{overlap_via, overlap_weight});
+                }
+                else
+                {
+                    shortest_path_via = overlap_via;
+                }
+            }
+        }
+
+        // Adjusting upper bound for forward search
+        shortest_path_weight = std::min(shortest_path_weight, overlap_weight);
+        // Force reverse step to not break early when we reached the middle, continue for overlap.
+        overlap_weight = INVALID_EDGE_WEIGHT;
+
+        if (!reverse_heap.Empty())
+        {
+            routingStep<REVERSE_DIRECTION>(facade,
+                                           reverse_heap,
+                                           forward_heap,
+                                           overlap_via,
+                                           overlap_weight,
+                                           force_loop_forward,
+                                           force_loop_backward,
+                                           phantom_node_pair);
+
+            if (!reverse_heap.Empty())
+                reverse_heap_min = reverse_heap.MinKey();
+
+            if (overlap_weight != INVALID_EDGE_WEIGHT && overlap_via != SPECIAL_NODEID)
+            {
+                if (shortest_path_via != SPECIAL_NODEID)
+                {
+                    candidate_vias.push_back(WeightedViaNode{overlap_via, overlap_weight});
+                }
+                else
+                {
+                    shortest_path_via = overlap_via;
+                }
+            }
+        }
+
+        // Adjusting upper bound for reverse search
+        shortest_path_weight = std::min(shortest_path_weight, overlap_weight);
+    }
+
+    const bool has_valid_shortest_path_weight = shortest_path_weight != INVALID_EDGE_WEIGHT;
+    const bool has_valid_shortest_path_via = shortest_path_via != SPECIAL_NODEID;
+    const bool has_shortest_path = has_valid_shortest_path_weight && has_valid_shortest_path_via;
+
+    if (!has_shortest_path)
+        return InternalManyRoutesResult{};
+
+    std::cout << ">>> shortest path weight: " << shortest_path_weight << std::endl;
+    std::cout << ">>> number of candidates: " << candidate_vias.size() << std::endl;
+
+    // Filters via candidate nodes with heuristics
+
+    // Note: filter pipeline below only makes range smaller; no need to erase items
+    // from the vector when we can mutate in-place and for filtering adjust iterators.
+    auto it = end(candidate_vias);
+
+    it = filterViaCandidatesByUniqueNodeIds(begin(candidate_vias), it);
```

Comment:
depending on the size, we could potentially do better with `n_th` element rankings, just as an idea for later optimisations.